### PR TITLE
CI/CD: add inclavare containers development centos8.2 and ubuntu 18.04 docker images

### PR DESCRIPTION
**Pull request by hustliyilin**
Signed-off-by: Yilin Li YiLin.Li@linux.alibaba.com

### CODE REVIEW

File: .github/workflows/docker/Dockerfile-dev-centos8.2 [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+RUN dnf --enablerepo=PowerTools install -y iptables && \
+    wget https://download.docker.com/linux/static/stable/x86_64/docker-18.09.7.tgz && \
+    tar -zxvf docker-18.09.7.tgz && mv docker/* /usr/bin
+
+RUN mkdir -p /etc/docker && \
+    echo -e "{\n\t\"runtimes\": {\n\t\t\"rune\": {\n\t\t\t\"path\": \"/usr/local/bin/rune\",\n\t\t\t\"runtimeArgs\": []\n\t\t}\n\t}\n}" >> /etc/docker/daemon.json
+
+# install dcap
+RUN wget https://download.01.org/intel-sgx/sgx-linux/2.13/distro/centos8.2-server/sgx_rpm_local_repo.tgz && \
+    tar -zxvf sgx_rpm_local_repo.tgz && rm sgx_rpm_local_repo.tgz && cd sgx_rpm_local_repo && rpm -ivh libsgx-headers*.rpm && rpm -ivh libsgx-dcap-quote-verify*.rpm && cd ../ && rm -rf sgx_rpm_local_repo
```
Comment (jiazhang0) [inline .github/workflows/docker/Dockerfile-dev-centos8.2:33]: I think we have talked this several times.

**Review by jiazhang0 — CHANGES_REQUESTED** [review]

File: .github/workflows/docker/Dockerfile-dev-centos8.2 (hunk 1 above)
Comment (hustliyilin) [reply .github/workflows/docker/Dockerfile-dev-centos8.2:33]: > I think we have talked this several times.

Ok, no problem.

File: tools/docker/Dockerfile-dev-ubuntu18.04 [hunk 2]
```diff
@@ -0,0 +1,36 @@
+FROM ubuntu:18.04
+
+LABEL maintainer="YiLin Li <YiLin.Li@linux.alibaba.com>"
+
+RUN apt-get update && apt-get install -y make wget git \
+    libseccomp-dev gnupg gcc pkg-config libssl-dev libprotobuf-c0-dev \
```
Comment (hustliyilin) [inline tools/docker/Dockerfile-dev-ubuntu18.04:6]: protobuf should be built from source codes.

**Comment by hustliyilin** [comment]
- build test: https://github.com/hustliyilin/inclavare-containers/runs/2022757822?check_suite_focus=true
- the build of rune、shim、 epm、sgx-tools、skeleton is ok.
- the docker is installed in the docker images.
- the rpm and deb build don't need to be installed any packages.

File: .github/workflows/manually_dev_image.yml [hunk 3]
```diff
@@ -0,0 +1,79 @@
+name: Build inclavare containers development images
```
Comment (jiazhang0) [inline .github/workflows/manually_dev_image.yml:1]: Rename to manual_inclavare_dev_image.yml

File: .github/workflows/manually_dev_image.yml [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        description: 'The version of inclavare containers image'
+        required: true
+
+jobs:
+  generate-centos-image:
+      # GitHub Actions doesn't have CentOS VM provided
+      runs-on: ubuntu-18.04
+
+      steps:
+      - name: Checkout code
```
Comment (jiazhang0) [inline .github/workflows/manually_dev_image.yml:16]: Checkout workspace.

File: .github/workflows/manually_dev_image.yml [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        uses: actions/checkout@v2
+
+      # Because "Build and push" step `context` field can't be subdir,
```
Comment (jiazhang0) [inline .github/workflows/manually_dev_image.yml:19]: Change to`the context field mentioned in the step "Build and push"`

File: .github/workflows/manually_dev_image.yml [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+      # we need to copy files needed by dockerfile to root dir of the project
+      - name: Copy context for docker build
+        run: |
+          cp -r .github/workflows/docker .
+
+      - name: Set up QEMU
```
Comment (jiazhang0) [inline .github/workflows/manually_dev_image.yml:25]: qemu is required for us?

File: tools/docker/Dockerfile-dev-centos8.2 [hunk 7]
```diff
@@ -0,0 +1,41 @@
+FROM centos:8.2.2004
```
Comment (jiazhang0) [inline tools/docker/Dockerfile-dev-centos8.2:1]: Please rename this dockerfile to Dockerfile-inclavare-dev-centos8.2

File: tools/docker/Dockerfile-dev-centos8.2 [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+LABEL maintainer="YiLin Li <YiLin.Li@linux.alibaba.com>"
+
+RUN dnf clean all && rm -r /var/cache/dnf && \
+    dnf --enablerepo=PowerTools install -y wget make libseccomp-devel \
+    gcc git openssl-devel glibc-static binutils-devel protobuf-c-devel \
+    rpm-build yum-utils
```
Comment (jiazhang0) [inline tools/docker/Dockerfile-dev-centos8.2:8]: The dependencies are not enough.

File: tools/docker/Dockerfile-dev-centos8.2 [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    echo -e "{\n\t\"runtimes\": {\n\t\t\"rune\": {\n\t\t\t\"path\": \"/usr/local/bin/rune\",\n\t\t\t\"runtimeArgs\": []\n\t\t}\n\t}\n}" >> /etc/docker/daemon.json
+
+# install SGX
+RUN [ ! -f sgx_rpm_local_repo.tgz ] && \
+    wget -c https://download.01.org/intel-sgx/latest/linux-latest/distro/centos8.2-server/sgx_rpm_local_repo.tgz && \
+    tar xzf sgx_rpm_local_repo.tgz && \
+    yum-config-manager --add-repo sgx_rpm_local_repo && \
+    yum makecache && rm sgx_rpm_local_repo.tgz
+
+RUN yum install --nogpgcheck -y libsgx-dcap-quote-verify-devel
```
Comment (jiazhang0) [inline tools/docker/Dockerfile-dev-centos8.2:41]: ditto

**Review by jiazhang0 — CHANGES_REQUESTED** [review]

**Comment by hustliyilin** [comment]
Done, thanks.

File: tools/docker/Dockerfile-dev-centos8.2 (hunk 8 above)
Comment (hustliyilin) [reply tools/docker/Dockerfile-dev-centos8.2]: > The dependencies are not enough.

Done, thanks.

File: tools/docker/Dockerfile-inclavare-dev-ubuntu18.04 [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    tar -zxvf docker-18.09.7.tgz && mv docker/* /usr/bin
+
+# configure the rune runtime of docker
+RUN mkdir -p /etc/docker && \
+    echo "{\n\t\"runtimes\": {\n\t\t\"rune\": {\n\t\t\t\"path\": \"/usr/local/bin/rune\",\n\t\t\t\"runtimeArgs\": []\n\t\t}\n\t}\n}" >> /etc/docker/daemon.json
+
+# install SGX
+RUN echo "deb [arch=amd64] https://download.01.org/intel-sgx/sgx_repo/ubuntu bionic main" | tee /etc/apt/sources.list.d/intel-sgx.list && wget -qO - https://download.01.org/intel-sgx/sgx_repo/ubuntu/intel-sgx-deb.key | apt-key add -
+
+RUN apt-get update -y && apt-get install -y libsgx-uae-service libsgx-dcap-quote-verify libsgx-dcap-quote-verify-dev
```
Comment (jiazhang0) [inline tools/docker/Dockerfile-inclavare-dev-ubuntu18.04:43]: libsgx-dcap-quote-verify is dependent by libsgx-dcap-quote-verify-dev so it is not necessary to explicitly write it.

File: tools/docker/Dockerfile-inclavare-dev-ubuntu18.04 [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+RUN apt-get install -y iptables && \
+    wget https://download.docker.com/linux/static/stable/x86_64/docker-18.09.7.tgz && \
+    tar -zxvf docker-18.09.7.tgz && mv docker/* /usr/bin
+
+# configure the rune runtime of docker
+RUN mkdir -p /etc/docker && \
+    echo "{\n\t\"runtimes\": {\n\t\t\"rune\": {\n\t\t\t\"path\": \"/usr/local/bin/rune\",\n\t\t\t\"runtimeArgs\": []\n\t\t}\n\t}\n}" >> /etc/docker/daemon.json
+
+# install SGX
+RUN echo "deb [arch=amd64] https://download.01.org/intel-sgx/sgx_repo/ubuntu bionic main" | tee /etc/apt/sources.list.d/intel-sgx.list && wget -qO - https://download.01.org/intel-sgx/sgx_repo/ubuntu/intel-sgx-deb.key | apt-key add -
```
Comment (jiazhang0) [inline tools/docker/Dockerfile-inclavare-dev-ubuntu18.04:41]: split into multi-lines.

**Review by jiazhang0 — CHANGES_REQUESTED** [review]

File: tools/docker/Dockerfile-inclavare-dev-ubuntu18.04 (hunk 10 above)
Comment (hustliyilin) [reply tools/docker/Dockerfile-inclavare-dev-ubuntu18.04:43]: > libsgx-dcap-quote-verify is dependent by libsgx-dcap-quote-verify-dev so it is not necessary to explicitly write it.

Ok

**Review by jiazhang0 — APPROVED** [review]

**Pull request merged**